str/associatr/pseudobulk.py

```diff
@@ -25,13 +26,19 @@
 from cpg_utils.hail_batch import get_batch, image_path, output_path
 
 
-def pseudobulk(input_file_path, target_sum, min_pct):
+def pseudobulk(input_file_path, id_file_path, target_sum, min_pct):
     """
     Performs pseudobulking in a cell-type chromosome-specific manner
     """
     expression_h5ad_path = to_path(input_file_path).copy('here.h5ad')
     adata = sc.read_h5ad(expression_h5ad_path)
 
+    # retain only samples in the id file
+    with to_path(id_file_path).open() as csvfile:
+        cpg_ids = list(csv.reader(csvfile))
+    cpg_ids = cpg_ids[0]  # the function above creates a list in a list
```

Comment:
if this creates a list in a list, why are you wrapping it in list? Isn't that the cause of the double-list?

Reply:
If I don't wrap it as a list it becomes a CSV reader object, that is not iterable.

Reply:
Reader objects are designed to be iterable... the following snippet works:
```
with to_path('gs://YOUR_INPUT.csv').open() as h:
    x = csv.reader(h)
    for y in x:
        print(y)
```

`y` here is a 1055 element long list, as your CSV entries are all on one line. If you were trying to access the first element of the results it wouldn't work, as reader objects aren't exactly lists:

```
x = csv.reader(h)
cpg_ids = x[0]
>>> TypeError: '_csv.reader' object is not subscriptable
```

A `csv.reader` object implements a `next` method which allows you to iterate through the underlying file one line at a time for scalability. By wrapping it in `list` you're forcing all those lines to be read at once and stored in memory. Here that's not an issue as all your elements are on one line anyway, but if your CSV was ~GB in size this would be a problem. This is a mini snippet for 'open the file, read the first line':

```
x = csv.reader(to_path('gs://YOUR_INPUT.csv').open())
cpg_ids = next(x)
```

None of this really matters, your implementation works, but it's good to know why it works.